For #15543 - Expand / collapse tabs tray depending on all tabs

Mugurell opened the pull request:
Now check all tabs - private or normal and decide whether to show the tabs tray
as expanded or collapsed depending on the highest number of tabs in any
category.
Also added support for the grid based tabs tray - will show tabs tray as
expanded if there are more than 3 or more tabs (normal or private) open
(this meaning two rows in the grid layout).

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] **Tests**:  No tests. Small changes in a not yet tested component.
- [x] **Screenshots**: Tried to record screen but with this involving private mode much of the video is blaked out.
- [x] **Accessibility**: The code in this PR or does not include any user facing features

| List based layout  | Grid based layout |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/11428869/96576343-b8f27500-12da-11eb-8e7a-ab9a453cb86b.gif" width="300" /> | <img src="https://user-images.githubusercontent.com/11428869/96576459-e3dcc900-12da-11eb-806a-accaf489342f.gif" width="300" /> |
| [video](https://drive.google.com/file/d/1_KzwLvbtcEphnSJdKZP89d9Xoc5FEp2q/view?usp=sharing) | [video](https://drive.google.com/file/d/1nGi1PDHlKeJFI0WrUGc-vGWNlcaJbfwY/view?usp=sharing) |

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

Mugurell commented:
Rebased after the grid layout changes for tabs tray.
Seems like the layout is still WIP so I didn't know if I should do anything to support it atm but tagging @gabrielluong for visibility of the changes that will need tp be supported also by the new layout.

gabrielluong commented:
> Rebased after the grid layout changes for tabs tray.
> Seems like the layout is still WIP so I didn't know if I should do anything to support it atm but tagging @gabrielluong for visibility of the changes that will need tp be supported also by the new layout.

Thanks for the ping! I will test this tomorrow morning and see how it interacts. I didn't see any UX feedback from @topotropic in the bugs mentioned, and an alarm kinda goes off internally for me when I see these kind of scenarios where we end up working on something prior to any UX investigation, which I always think becomes an awkward scenario for all of us. Would you be able to ping her to get some feedback? I could also be mistaken and that these conversations already took place, and if that is the case, let me apologize in advance.

Mugurell commented:
> Thanks for the ping! I will test this tomorrow morning and see how it interacts. I didn't see any UX feedback from @topotropic in the bugs mentioned, and an alarm kinda goes off internally for me when I see these kind of scenarios where we end up working on something prior to any UX investigation, which I always think becomes an awkward scenario for all of us. Would you be able to ping her to get some feedback? I could also be mistaken and that these conversations already took place, and if that is the case, let me apologize in advance.

This patch just comes to complement what we're already doing when the tab tray first shows up.
It can take the entire screen or just half depending on how many tabs there are to show.
But we didn't account for different tab tray sizes, expanded tabs tray needed for normal tabs and collapsed tray needed for private tabs for example.
Will ask for UX feedback in the ticket.

gabrielluong commented:
@topotropic Would you be able to give this a try and provide any feedback? I personally think it looks good.

gabrielluong commented:
I think it's also somewhat safe to say the grid layout is in a somewhat stable state. We are only working to align it more with the designs (something that I am not really good at).

Mugurell commented:
> I think it's also somewhat safe to say the grid layout is in a somewhat stable state. We are only working to align it more with the designs (something that I am not really good at).

For showing the tab tray as collapsed / expanded we just counted the number of tabs to be displayed.
In the list based layout the minimum number of tabs for which we'll show the tab tray is 4.
Maybe for the grid based layout we should show tabs tray as expanded if there are 3 or more tabs to be shown?
Guess these are questions for @topotropic also.
With >=2 this is how it'd look.
<img src="https://user-images.githubusercontent.com/11428869/95959466-2bf77980-0e0b-11eb-81e8-93a9d7915b3e.png" width="300">

I think we could also look into using another more suited height for the grid based collapsed tabs tray but if the changes are bigger I'd say it would be best to have a new ticket specifically for this.

gabrielluong commented:
@Mugurell let's add a gif for @topotropic so she can review it quicker although my experience with trying out this PR is that it's best to try it with an apk.

Mugurell commented:
> @Mugurell let's add a gif for @topotropic so she can review it quicker although my experience with trying out this PR is that it's best to try it with an apk.

Updated the patch to set tabs tray as expanded / collapsed only when first showing on the screen, not also when changing what tabs to show.
Added gif / video for the behaviors of both the list based and grid based tabs trays.
An apk on which these changes can be tested is available from the CI tasks - https://firefox-ci-tc.services.mozilla.com/tasks/ekFcvlBQSU-FSGPPXlbTQw#artifacts

topotropic reviewed (approved):
LGTM, thanks!

gabrielluong commented on app/src/main/java/org/mozilla/fenix/tabtray/TabTrayView.kt:
```diff
@@ -280,6 +275,30 @@ class TabTrayView(
+    private fun getTabsNumberInAnyMode(): Int {
+        val normalTabsNumber = view.context.components.core.store.state.normalTabs.size
+        val privateTabsNumber = view.context.components.core.store.state.privateTabs.size
+        return if (normalTabsNumber > privateTabsNumber) {
```
We can just return the max()

Mugurell replied:
Thanks!

gabrielluong commented on app/src/main/java/org/mozilla/fenix/tabtray/TabTrayView.kt:
```diff
@@ -312,8 +331,12 @@ class TabTrayView(
-    fun expand() {
-        behavior.state = BottomSheetBehavior.STATE_EXPANDED
+    fun accomodateTabs() {
+        if (getTabsNumberInAnyMode() >= getTabsNumberForExpandingTray() || isInLandscape()) {
```
Let's flip the order of the `||` and have `isInLandscape()` first to allow the if statement to escape quicker if that one condition is true.

Mugurell replied:
Thanks! Will change it.
I was going with the presumption that most of the time the device will be in portrait but have not data to back this up.

Mugurell commented:
I see `test-debug` failing because of `[taskcluster:error] Task timeout after 7200 seconds. Force killing container.`
Restarting CI tasks.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/tabtray/TabTrayView.kt:
```diff
@@ -145,20 +146,14 @@ class TabTrayView(
-        if (tabs.size > EXPAND_AT_SIZE || startingInLandscape) {
-            expand()
-        }
+        accomodateTabs()
```
Let's rename this to `updateBottomSheetBehavior()`

Mugurell replied:
I feel like with this naming we're offering implementation details that the caller should not be concerned about while still not clearly expressing the intent. The kdoc suggested below will help with that.

gabrielluong replied:
I am thinking about this from the perspective of just reading the code without jumping to the function. This would be absolutely clear to me without jumping and looking at what `accomodateTabs()` if we added (for example):

```
// Adjust the bottom sheet height to accommodate the number of tabs to display.
updateBottomSheetBehavior()
```

Alternative, naming would be `updateTabTrayHeight`, `adjustTabTrayHeight` or some variant of that. I could see these namings without the comment needed.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/tabtray/TabTrayView.kt:
```diff
@@ -312,8 +331,12 @@ class TabTrayView(
 
-    fun expand() {
-        behavior.state = BottomSheetBehavior.STATE_EXPANDED
+    fun accomodateTabs() {
```
Let's add a kDoc for this. I want to encourage clearer comments for functionality overall.

```
/**
 * Updates the bottom sheet visibility based on the number tabs or screen orientation.
 * Show the bottom sheet fully expanded if it is in landscape mode or the number of 
 * tabs are greater or equal to the expand size limit.
 */
```

Mugurell replied:
I don't see the intent of this as to update the visibility (as in visible / invisible) but rather the height of the bottom sheet.
Could I change the comment to
```
/**
 * Updates the bottom sheet height based on the number tabs or screen orientation.
 * Show the bottom sheet fully expanded if it is in landscape mode or the number of 
 * tabs are greater or equal to the expand size limit.
 */
```
?

gabrielluong replied:
Sounds good

gabrielluong commented on app/src/main/java/org/mozilla/fenix/tabtray/TabTrayView.kt:
```diff
@@ -666,7 +689,8 @@ class TabTrayView(
         private const val DEFAULT_TAB_ID = 0
         private const val PRIVATE_TAB_ID = 1
-        private const val EXPAND_AT_SIZE = 3
+        private const val EXPAND_AT_LIST_SIZE = 4
```
Let's add some comments for these 2 constants too.

`
// Minimum number of grid tabs to trigger a fully expanded bottom sheet when opening the tab tray.
`

Mugurell replied:
Is it ok to use ```// Minimum number of grid tabs for which to show the tabs tray as expanded.```?

gabrielluong replied:
Ya sounds good.

gabrielluong reviewed (approved):
r+ assuming we make the changes.

Mugurell commented on app/src/main/java/org/mozilla/fenix/tabtray/TabTrayDialogFragment.kt:
```diff
@@ -161,14 +161,13 @@ class TabTrayDialogFragment : AppCompatDialogFragment(), UserInteractionHandler
                 // orientation changes.
                 tabTrayView.updateTabsTrayLayout()
             }
-
```
Wasn't really sure why this was here.
Normally I'm a big believer in not modifying code unneededly. Makes the future archaeological work a lot harder.
Here I see ktlint complaining about `Unexpected blank line` so... I think the rules were modified in the meantime and we have to now remove this newline.
